assert: fix boxed primitives in deep*Equal

BridgeAR opened the pull request:
Unbox all primitives and compare them as well instead of
only comparing boxed strings.

<s>This is blocked by #15036 because there is a test case from N-API that currently only passes because deepStrictEqual is not testing them properly and this hides the underlying NaN issue.
```
AssertionError [ERR_ASSERTION]: [Number: NaN] deepStrictEqual [Number: NaN]
    at Object.<anonymous> (/node/test/addons-napi/test_conversions/test.js:119:8)
```
</s>

Update: I commented out the failing test for now.
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert

kunalspathak commented on lib/assert.js:
```diff
@@ -203,6 +203,11 @@ function strictDeepEqual(actual, expected) {
       return true;
     }
+  } else if (typeof actual.valueOf === 'function') {
+    const rawA = actual.valueOf();
```
>rawA [](start = 10, length = 4)

nit: can you have better name like `rawAcual` ?

BridgeAR commented:
I rebased due to conflicts, addressed the nit and added a fast path for boxed primitives. I also commented out the failing test, so now it is not blocked anymore.

jasnell commented:
Curious about the semver-iness of this change. Thoughts @nodejs/tsc?

BridgeAR commented:
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/961/

BridgeAR commented:
Rebased due to conflicts.

BridgeAR commented:
Ping @nodejs/ctc @nodejs/tsc PTAL

mcollina commented on test/addons-napi/test_conversions/test.js:
```diff
@@ -116,7 +116,8 @@ assert.deepStrictEqual(new Boolean(false), test.toObject(false));
 assert.deepStrictEqual(new Number(0), test.toObject(0));
-assert.deepStrictEqual(new Number(Number.NaN), test.toObject(Number.NaN));
+// TODO: Add test back in as soon as NaN is properly compared
+// assert.deepStrictEqual(new Number(Number.NaN), test.toObject(Number.NaN));
```
Can we remove the todo and add the test back?

BridgeAR replied:
Not before #15036 landed. Otherwise the test would fail.

hthomison1078 replied:
#15036

hthomison1078 replied:
Killswitch

hthomison1078 replied:
Don't Forget

mcollina replied:
what's the depedency of this change to https://github.com/nodejs/node/pull/15036?

BridgeAR replied:
Currently `assert.deepStrictEqual(NaN, NaN)` will fail. This test only worked because `NaN` was never compared as it was not unboxed before. #15036 will change that to not fail anymore in case `NaN` is compared to itself.

joyeecheung replied:
Have you pushed you changes to this PR? Looks like these lines are still commented?

BridgeAR replied:
Thanks a lot for reminding me! It is now back in place.

joyeecheung commented on lib/assert.js:
```diff
@@ -210,13 +210,30 @@ function strictDeepEqual(actual, expected) {
         Object.keys(expected).length === expected.length) {
       return true;
     }
+  } else if (typeof actual.valueOf === 'function') {
```
IIUC this also affects objects with custom `valueOf`, so the documentation needs to be updated as well..

BridgeAR replied:
You are right! I am going to add that to the documentation.

BridgeAR replied:
Addressed

Trott commented on doc/api/assert.md:
```diff
@@ -161,6 +162,11 @@ assert.deepEqual(date, fakeDate);
+assert.deepStrictEqual({}, new Number(1));
+// Fails as the wrapped number is unwrapped and compared as well.
+assert.deepStrictEqual(new String('foo'), Object('foo'));
+// Ok, both objects and strings are unwrapped identical.
```
Nit: Perhaps change the comment to something like:

> OK because the object and the string are identical when unwrapped.

Trott replied:
I'm confused by this documentation. It's supposed to document new behavior you're adding, right? But the assertions here are already fail/succeed as described in Node.js 8.4.0 so the behavior isn't because of any changes added here. Am I misunderstanding?

BridgeAR replied:
Changed. I indeed chose a bad example for the number and updated that. Currently only wrapped strings are compared but in my former example with the wrapped number, the `toString()` was not identical and it failed early.

Trott commented on doc/api/assert.md:
```diff
@@ -161,6 +162,11 @@ assert.deepEqual(date, fakeDate);
 // Different type tags
+
+assert.deepStrictEqual({}, new Number(1));
+// Fails as the wrapped number is unwrapped and compared as well.
```
Micro-nit: `Fails as the wrapped number is...` -> `Fails because the wrapped number is...`

BridgeAR replied:
Thanks for the suggestion!

BridgeAR commented:
Rebased

joyeecheung commented:
LGTM with the commented test uncommented

jasnell commented:
this would need to be backported for v8.x
